File: p037.py
from prime_sieve import PrimeSieve

def get_truncations(n):
	yield n
	d = 10
	while d <= n:
		yield n % d
		yield n // d
		d *= 10

def get_truncatable_primes():
	till = 1000000
	sieve = PrimeSieve(till)
	for n in range(11, till, 2):
		if all(sieve.is_prime(m) for m in get_truncations(n)):
			yield n

if __name__ == "__main__":
	truncatable_primes = get_truncatable_primes()
	solution = sum(truncatable_primes)
	print(solution)

File: p037.rs
mod prime_sieve;

fn is_truncable_prime(prime_sieve: &prime_sieve::PrimeSieve, n: usize) -> bool {
    if !prime_sieve.is_prime(n) { return false; }
    let mut power_of_10 = 10;
    while power_of_10 <= n {
        if !prime_sieve.is_prime(n/ power_of_10) { return false; }
        if !prime_sieve.is_prime(n% power_of_10) { return false; }
        power_of_10 *= 10;
    }
    true
}

fn main() {
    const EXCLUDED_UPPER_BOUND: usize = 1_000_000;
    let prime_sieve = prime_sieve::PrimeSieve::new(EXCLUDED_UPPER_BOUND);

    let solution: usize = (11..EXCLUDED_UPPER_BOUND)
        .step_by(2)
        .filter(|&n| is_truncable_prime(&prime_sieve, n))
        .sum();

    println!("{}", solution);
}